### DISCUSSION TITLE
federation router, schema first and federation subscirpitoins questions :)

### DISCUSSION THREAD

**Discussion by guykatz** [smallrye/smallrye-graphql#2178] [Q&A] [answered]
HI all; I have a monolitic graphql-java implementation (schema first, not using smallrye gql).
I am trying to navigate the federation landscape and I have a few questions:
1. can smallrye gql be used as a router? or is it just for the sub graphs. (if not are there open source routers in java?)
2. is smallrye gql still only code first and not schema first?
3. does federation support subsciptions? (not a smallrye specific question but I am here anyway so maybe someone can help)
thanks!

**Comment by jmartisk** [comment]
Hi, 
I'm not very knowledgeable in federation in particular, but I'll try to answer what I can
1) no it's just for subgraphs. I haven't heard of a Java-based router yet (not saying for sure that there isn't one)
2) yes, only code first. We could implement some code generator that generates GraphQL API from a schema definition, but it's not considered a high priority.. (community contributions are welcome)
3) AFAIK not natively, only through some workarounds, but maybe somebody more knowledgeable about Federation can help (@RoMiRoSSaN, @RoKKim)

**Comment by t1** [comment]
> 1. can smallrye gql be used as a router? or is it just for the sub graphs. (if not are there open source routers in java?)

It's currently only for the subgraphs, and I'm not aware of any GraphQL Federation Router written in Java. @phillip-kruger and I started [Feder](https://github.com/GraphQL-Feder/feder), but we didn't find the time to push it forward from the initial, PoC state.

But why would you base your choice of a router on the programming language? It could be helpful for a migration, when you extract GraphQL code without Federation support into separate Federation enabled services. But when done, you should be free to switch to any other Router.

> 2. is smallrye gql still only code first and not schema first?

Yes. I'm sure you are aware, that this a long discussion 😁. I'm personally not the biggest fan of the idea to generate code from a schema, probably because I did too much SOAP in the past 🤪 And writing the schema _and_ the code by hand, is not very productive. I prefer to define my domain logic, and then just sprinkle it with a pinch of annotation fairy dust.

If you take client-first seriously, I'd go all the way and go for Consumer Driven Contracts (maybe even with [WunderBar](https://github.com/t1/wunderbar)... sorry for pitching my own stuff again).

> 3. does federation support subsciptions?

I've never used it, yet, but if the Router supports it, it can even federate the types from the substription. See [this example](https://www.apollographql.com/docs/router/executing-operations/subscription-support/#example-execution)

**Reply by guykatz** [reply]
@t1  @jmartisk thanks for the answers. 
- you guys are right, I do not want to base my decision on java but a java router will probably feel at home with debuging etc but what are the most widely used, open source, federation routers out there?
- as for schema first, i was wondering, form your experience, how easy is it to migrate to code first? my only restriction is that migrating must be 1-1 similar so the schema produced by the code first must be the same as my current schema... we use most of the constructs that the currect schema spec allows (which is not that many)
- as for subscriptions, our solution relies heavily on them so i assume i will need a federated router that supports them. 
thanks again.

**Comment by RoMiRoSSaN** [comment, marked as answer]
1. Personally, I came across a router for graphql written in java. But it was not fully completed, had bugs, and did not support federations (it could only unite completely different subgraphs). I don’t think there is a production ready solution in Java, but I could be wrong. If you need a completely open source routing solution, I might suggest wundergraph cosmo (used in production). Quite a mature solution, but still evolving. There is also, for example, the guild hive, which is also not a bad solution.

2. Only code first. It may not be a familiar transition, but in many places it's worth it. Although in the federation, if you do very strange things, it can be painful, you will have to figure it out. You need to try this). There is a request from my team that it would be nice to generate code in some cases. There is a [plugin](https://github.com/kobylynskyi/graphql-java-codegen). We use it in some projects. I plan to see in the near future whether this can be adapted for 'smallrye like code' generation with minimal effort.

3. Subscriptions are supported. As for the router, you need to look directly at the documentation of a particular solution. Apollo, Cosmo, Hive support. True, at Apollo it seems to be paid. I know that this works for sure on cosmo, and they also offer alternative subscription methods, for example through kafka/nats. If you meant whether a federated object can be collected when using subscriptions, then I just checked on cosmo, yes it works there.
